### PR TITLE
Medium, low and QA tweaks

### PR DESCRIPTION
**Pull request by jefflau**
1. switch to `safeTransferFrom` https://github.com/code-423n4/2022-07-ens-findings/issues/156
2. Fix natspec in ReverseRegistrar.sol
3. Add `DefaultReverseResolverChanged` event in ReverseRegistrar.sol
4. Remove unused CustomError from Namewrapper.sol
5. Fix NatSpec on Namewrapper.sol
6. Move `_setData` to other internal functions section
7. Add IERC165 to ETHRegistrarController.sol
8. Custom Errors for controller
9. Add ERC20Recoverable contract and added to Controller/NameWrapper
10. Switched to named imports for clarity

### CODE REVIEW

File: contracts/utils/FundsRecoverable.sol [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+contract FundsRecoverable is Ownable {
+    /**
+    @notice Recover ERC20 tokens sent to the contract by mistake.
+    @dev The contract is Ownable and only the owner can call the recover function.
+    @param _to The address to send the tokens to.
+    @param _token The address of the ERC20 token to recover. 0x0 for ETH.
+    @param _amount The amount of tokens to recover.
+ */
+    function recoverFunds(
```
Comment (makoto) [inline contracts/utils/FundsRecoverable.sol:19]: Test?

File: contracts/ethregistrar/BulkRenewal.sol [hunk 2]
```diff
@@ -1,5 +1,5 @@
-pragma solidity >=0.8.4;
-pragma experimental ABIEncoderV2;
+//SPDX-License-Identifier: MIT
+pragma solidity >=0.8.17 <0.9.0;
```
Comment (Arachnid) [inline contracts/ethregistrar/BulkRenewal.sol:2]: This should be equivalent to `pragma solidity ~0.8.17`.

File: contracts/ethregistrar/BulkRenewal.sol [hunk 3]
```diff
@@ -53,18 +57,22 @@ contract BulkRenewal is IBulkRenewal {
...
+            uint256 totalPrice = price.base + price.premium;
+            controller.renew{value: totalPrice}(names[i], duration);
+            unchecked {
+                ++i;
+                total += totalPrice;
+            }
         }
         // Send any excess funds back
-        payable(msg.sender).transfer(address(this).balance);
+        payable(msg.sender).transfer(msg.value - total);
```
Comment (Arachnid) [inline contracts/ethregistrar/BulkRenewal.sol:75]: Doesn't this ensure that any ETH sent accidentally is just stuck forever?

File: contracts/ethregistrar/ETHRegistrarController.sol [hunk 4]
```diff
@@ -1,20 +1,38 @@
...
+import {ReverseRegistrar} from "../registry/ReverseRegistrar.sol";
+import {IETHRegistrarController, IPriceOracle} from "./IETHRegistrarController.sol";
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+import {INameWrapper} from "../wrapper/INameWrapper.sol";
+import {ERC20Recoverable} from "../utils/ERC20Recoverable.sol";
+
+error CommitmentTooYoung(bytes32 commitment);
```
Comment (Arachnid) [inline contracts/ethregistrar/ETHRegistrarController.sol:16]:
```suggestion
error CommitmentTooNew(bytes32 commitment);
```

**Review by Arachnid — APPROVED** [review]

File: contracts/ethregistrar/BulkRenewal.sol (hunk 3 above)
Comment (jefflau) [reply contracts/ethregistrar/BulkRenewal.sol:75]: Is it possible to send ETH here accidentally? I thought it reverts since there is no receive function.

**Review by Arachnid — APPROVED** [review]

**Pull request merged**